exercises/book/A3-traits-generics/mod.md

```diff
@@ -0,0 +1,151 @@
+        todo!("Implement me");
+    }
+}
+```
+
+**Question** 
+1. How would you pronounce the first line of the code you just copied in English?*
+
+Run `cargo test` to validate your implementation.
+
+### A3.2.C `impl LocalStorageVec`
+To make the `LocalStorageVec` more useful, we'll add more methods to it. Create an `impl`-block for `LocalStorageVec`. Don't forget to declare and provide the generic paramereters. For now, to make implementations easier, we will add a bound `T`, requiring that it implements `Copy` and `Default`. First off, uncomment the test called `it_constructs`. Make it compile and pass by creating a assosiated function called `new` on `LocalStorageVec` that creates a new, empty `LocalStorageVec` instance without heap allocation.
+
+The next methods we'll implement are `len`, `push`, `pop`, `insert`, `remove` and `clear`:
+- `len` returns the length of the `LocalStorageVec`
+- `push` appends an item to the end of the `LocalStorageVec` and increments its length. Possibly moves the contents to the heap if they no longer fit on the stack.
+- `pop` removes an item from the end of the `LocalStorageVec`, optionally returns it and decrements its length. If the length is 0, `pop` returns `None`
+- `insert` inserts an item at the given index and increments the length of the `LocalStorageVec`
+- `remove` removes an item at the given index and returns it.
+- `clear` resets the length of the `LocalStorageVec` to 0.
+
+ Uncomment the corresponding test cases and make them compile and pass. **Be sure to have a look at the methods provided for slices [`[T]`](https://doc.rust-lang.org/std/primitive.slice.html) and [`Vec<T>`](https://doc.rust-lang.org/std/vec/struct.Vec.html)** Specifically, `[T]::copy_within` and `Vec::extend_from_slice` can be of use.
+
+
+### A3.2.D `Iterator` and `IntoIterator`
+Our `LocalStorageVec` can be used in the real world now, but we still shoudn't be statisfied. There are various traits in the standard library that we can implement for our `LocalStorageVec` that would make users of our crate happy. 
+
+First off, we will implement the [`IntoIterator`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html) and [`Iterator`](https://doc.rust-lang.org/std/iter/trait.Iterator.html) traits. Go ahead and uncomment the `it_iters` test case. Let's define a new type:
+
+
+```rust
+pub struct LocalStorageVecIter<T, const N: usize> {
+    vec: LocalStorageVec<T, N>,
+    counter: usize,
+}
+```
+
+This is the type we'll implement the `Iterator` trait on. You'll need to specify the item this `Iterator` implementation yields, as well as an implementation for `Iterator::next`, which yields the next item. You'll be able to make this easier by bounding `T` to `Default` when implementing the `Iterator` trait, as then you can use the [`std::mem::take`](https://doc.rust-lang.org/std/mem/fn.take.html) function to take an item from the `LocalStorageVec` and replace it with the default value for `T`.
+
```

Comment:
I have not needed `std::mem::take` in my solution so I don't understand this hint. Also, since at this point in the exercise the `impl` block for `LocalStorageVec` requires `T` to be `Default + Copy`, students are most likely to find they need the same constraint on the `LocalStorageVecIter` anyway, so I think the explanation here will only be confusing.